src/Features/CSharp/Portable/CodeFixes/AddParameterXmlDocumentation/AddParameterXmlDocumentationCodeFixProvider.cs

```diff
@@ -0,0 +1,316 @@
+                elementKind, out adjacentNameAttribute, out adjacentAttributeIsBefore);
+        }
+
+        private class FindAdjacentParameterDocumentationWalker : CSharpSyntaxWalker
+        {
+            private FindAdjacentParameterDocumentationWalker(int parameterIndex, Dictionary<string, int> parameterIndices, XmlNameAttributeElementKind elementKind)
+                : base(SyntaxWalkerDepth.StructuredTrivia)
+            {
+                _parameterIndex = parameterIndex;
+                _parameterIndices = parameterIndices;
+                _elementKind = elementKind;
+            }
+
+            private int _parameterIndex = -1;
+            private Dictionary<string, int> _parameterIndices = new Dictionary<string, int>();
+
+            private XmlNameAttributeSyntax _documentedParameterBefore = null;
+            private int _documentedParameterBeforeIndex = -1;
+            private XmlNameAttributeSyntax _documentedParameterAfter = null;
+            private int _documentedParameterAfterIndex = int.MaxValue;
+
+            private XmlNameAttributeElementKind _elementKind;
+
+            public static void FindAdjacentParameterDocumenteden(string parameterName, IEnumerable<string> allParameterNames,
+                SyntaxToken tokenWithDocumentationComment, XmlNameAttributeElementKind elementKind, out XmlNameAttributeSyntax adjacentNameAttribute, out bool adjacentAttributeIsBefore)
+            {
+                var parameterIndices = new Dictionary<string, int>();
+                int index = 0;
+                foreach (var param in allParameterNames)
+                {
+                    parameterIndices.Add(param, index);
+                    index++;
+                }
+                if (!parameterIndices.ContainsKey(parameterName))
+                {
+                    // something went very wrong
+                    Debug.Assert(false); // TODO message or remove assert
+                    adjacentNameAttribute = null;
+                    adjacentAttributeIsBefore = false;
+                    return;
+                }
+
+                var walker = new FindAdjacentParameterDocumentationWalker(parameterIndices[parameterName], parameterIndices, elementKind);
+                walker.VisitLeadingTrivia(tokenWithDocumentationComment);
+
+                if (walker._documentedParameterBefore != null)
+                {
+                    adjacentNameAttribute = walker._documentedParameterBefore;
+                    adjacentAttributeIsBefore = true;
+                }
+                else
+                {
+                    adjacentNameAttribute = walker._documentedParameterAfter;
+                    adjacentAttributeIsBefore = false;
+                }
+            }
+
+            public override void VisitXmlNameAttribute(XmlNameAttributeSyntax node)
+            {
+                if (node.GetElementKind() == _elementKind)
```

Comment:
FYI -- this doesn't look like a response to my comment above about the API change. Did you put it in the wrong place. I'm assuming this is in response to my comment about using a walker.

FWIW, I still think a walker is overkill to gather up `<param/>` tags that are poorly nested. It's easier to just use `DescendentNodes(...)` like so:

``` C#
var paramAttributes = myXmlDocComment
    .DescendentNodes().OfType<XmlNameAttributeSyntax>()
    .Where(n => string.Compare(n.Name.LocalName.ValueText, "param", StringComparison.OrdinalIgnoreCase))
```

FWIW, I would not even bother will esoteric malformed cases like the one you showed above. I've never seen someone intentionally write an XML doc comment that way.